src/providers/wms/qgswmsprovider.cpp

```diff
@@ -4486,19 +4607,29 @@ void QgsWmsTiledImageDownloadHandler::tileReplyFinished()
     // only take results from current request number
     if ( mTileReqNo == tileReqNo )
     {
-      double cr = mViewExtent.width() / mImage->width();
-
-      QRectF dst( ( r.left() - mViewExtent.xMinimum() ) / cr,
-                  ( mViewExtent.yMaximum() - r.bottom() ) / cr,
-                  r.width() / cr,
-                  r.height() / cr );
-
       QgsDebugMsgLevel( QStringLiteral( "tile reply: length %1" ).arg( reply->bytesAvailable() ), 2 );
 
       QImage myLocalImage = QImage::fromData( reply->readAll() );
 
       if ( !myLocalImage.isNull() )
       {
+        if ( mResamplingEnabled && mSourceResolution < 0 )
+        {
+          mSourceResolution = r.width() / myLocalImage.width();
+          int pixelX = std::ceil( mViewExtent.width() / mSourceResolution ) + 5;
+          int pixelY = std::ceil( mViewExtent.height() / mSourceResolution ) + 5;
+
+          *mImage = QImage( pixelX, pixelY, QImage::Format_ARGB32 );
+          mImage->fill( 0 );
+
+          mEffectiveViewExtent = QgsRectangle( mViewExtent.xMinimum() - 2 * mSourceResolution,
+                                               mViewExtent.yMinimum() - 2 * mSourceResolution,
+                                               mViewExtent.xMinimum() + ( mImage->width() - 2 ) * mSourceResolution,
+                                               mViewExtent.yMinimum() + ( mImage->height() - 2 ) * mSourceResolution );
+        }
```

Comment:
this code looks like a copy of the same code added earlier in the file - any chance to de-duplicate the logic?

also, it still uses 5 extra pixels instead of 4 :thinking: